include/pybind11/detail/type_caster_base.h

```diff
@@ -855,14 +911,21 @@ struct is_copy_constructible<std::pair<T1, T2>>
 template <typename T, typename SFINAE = void>
 struct is_copy_assignable : std::is_copy_assignable<T> {};
 template <typename Container>
-struct is_copy_assignable<Container,
-                          enable_if_t<all_of<std::is_copy_assignable<Container>,
-                                             std::is_same<typename Container::value_type &,
-                                                          typename Container::reference>>::value>>
+struct is_copy_assignable<
+    Container,
+    enable_if_t<
+        all_of<std::is_copy_assignable<Container>,
+               std::is_same<typename Container::value_type &, typename Container::reference>,
+               // Avoid infinite recursion
+               negation<is_recursive_container<Container>>>::value>>
     : is_copy_assignable<typename Container::value_type> {};
 template <typename T1, typename T2>
 struct is_copy_assignable<std::pair<T1, T2>>
-    : all_of<is_copy_assignable<T1>, is_copy_assignable<T2>> {};
+    /*
+     * Need to remove the const qualifier from T1 here since the value_type in
+     * STL map types is std::pair<const Key, T>, and const types are never assignable
+     */
+    : all_of<is_copy_assignable<typename std::remove_const<T1>::type>, is_copy_assignable<T2>> {};
```

Comment:
The asymmetry here seems odd, did you already think this through?

Concretely, would it make sense to apply `std::remove_const` also to `T2`?

Because for `std::pair` by itself (not considering how it's used elsewhere), the order of `T1` and `T2` doesn't have a function AFAIK.

Which leads me to think: It has to be either none or both.

Or we actually have a trait with subtly different semantics compared to the "pure" `is_copy_assignable`.

Reply:
Agreed on this point, the asymmetry is a bit odd.

Reply:
PLEASE DO NOT REVIEW YET, I need to go over this on my own again first.

Fixing this required changing the logic of this PR quite a bit.
The trouble of `is_recursive_container<T>` is that it only tells you that the type is recursive at all and then tells you to deal with it somehow. What I did now instead was replacing that with a `recursive_container_traits` that has a dependent type `recursive_container_traits<Container>::type_to_check_recursively`:

```cpp
/*
 * Tag to be used for representing the bottom of recursively defined types.
 * Define this tag so we don't have to use void.
 */
struct recursive_bottom {};

template <typename Container, typename SFINAE = void>
struct impl_recursive_container_traits {
    using type_to_check_recursively = recursive_bottom;
};

// and its main logic in the specialization
```

The `type_to_check_recursively` is defined such:

1. By default, it is `recursive_bottom`, so that the recursion is canceled.
2. If the type is non-recursive and defines a `value_type`, then the `value_type` is used. If the `value_type is a pair, the `const` is removed from the first type.
3. If the type is recursive and `value_type` is not a pair, then `recursive_bottom` is returned.
4. If the type is recursive and `value_type` is a pair, then `const` is removed from the first type and the first type is returned.

Having this trait makes the definitions of custom traits such as `is_move_constructible` very straightforward and no longer requires SFINAE constructs there:

```cpp
template <typename T>
struct is_move_constructible
    : all_of<std::is_move_constructible<T>,
             is_move_constructible<
                 typename recursive_container_traits<T>::type_to_check_recursively>> {};

template <>
struct is_move_constructible<recursive_bottom> : std::true_type {};

// Likewise for std::pair
// (after C++17 it is mandatory that the move constructor not exist when the two types aren't
// themselves move constructible, but this can not be relied upon when T1 or T2 are themselves
// containers).
template <typename T1, typename T2>
struct is_move_constructible<std::pair<T1, T2>>
    : all_of<is_move_constructible<T1>, is_move_constructible<T2>> {};
```

Reply:
> Fixing this required changing the logic of this PR quite a bit.

Oh, wow. Thanks a lot for going the extra mile!